core/src/main/scala/com/salesforce/op/stages/impl/insights/RecordInsightsLOCO.scala

```diff
@@ -135,29 +151,75 @@ class RecordInsightsLOCO[T <: Model[T]]
+      val (oldInd, oldVal) = featureArray(i)
+      val diffToExamine = computeDiffs(i, oldInd, oldVal, featureArray, featureSize, baseScore)
+      val max = diffToExamine(indexToExamine)
+
+      // Let's check the indicator value and descriptor value
+      val history = histories(oldInd)
+      val parentStages = history.parentFeatureStages
+      val indValue = history.indicatorValue
+      val descValue = history.descriptorValue
+
+      // If those values are empty, the field is likely to be a derived text feature (hashing tf output)
+      if (indValue.isEmpty && descValue.isEmpty) {
+        // Name of the field
+        val rawName = parentStages match {
+          case s if s.exists(_.contains(smartTextClassName)) => Some(history.parentFeatureOrigins.head)
+          case s if s.exists(_.contains(smartTextMapClassName)) => history.grouping
+          case _ => None
+        }
+        rawName.map { name =>
+          // Update the aggregation map
+          val (indices, array) = aggregationMap.getOrElse(name, (Array.empty[Int], Array.empty[Double]))
+          aggregationMap(name) = (indices :+ i, sumArray(array, diffToExamine))
+        }
+      } else {
+        addToHeaps(i, max, diffToExamine)
+      }
       i += 1
     }
+    // Adding LOCO results from aggregation map into heaps
+    aggregationMap.foreach {
+      case (_, (indices, ar)) => {
+        val i = indices.head // The index here is arbitrary
+        val n = indices.length
+        val diffToExamine = ar.map(_ / n)
+        val max = diffToExamine(indexToExamine)
+        addToHeaps(i, max, diffToExamine)
+      }
+    }
+
     val topPositive = positiveMaxHeap.dequeueAll
     val topNegative = negativeMaxHeap.dequeueAll
-    (topPositive ++ topNegative)
+    topPositive ++ topNegative
   }
 
-  override def transformFn: OPVector => TextMap = (features) => {
+  override def transformFn: OPVector => TextMap = features => {
+
     val baseResult = modelApply(labelDummy, features)
     val baseScore = baseResult.score
 
     // TODO sparse implementation only works if changing values to zero - use dense vector to test effect of zeros
     val featuresSparse = features.value.toSparse
-    val featureArray = featuresSparse.indices.zip(featuresSparse.values)
-    val filledSize = featureArray.length
+    // Besides non 0 values, we want to check the text features as well
+    val indices = (featuresSparse.indices ++ textIndices).toSet
+    val filledSize = indices.size
+    val featureArray = indices.map(i => i -> features.value(i)).toArray
     val featureSize = featuresSparse.size
 
     val k = $(topK)
+    // Index where to examine the difference in the prediction vector
     val indexToExamine = baseScore.length match {
       case 0 => throw new RuntimeException("model does not produce scores for insights")
       case 1 => 0
       case 2 => 1
-      case n if (n > 2) => baseResult.prediction.toInt
+      case n if n > 2 => baseResult.prediction.toInt
```

Comment:
I am not following - why prediction value becomes an index?!

Reply:
We want to return the LOCO score of the predicted class.

Let's say for a row the LOCOs are `0 -> LOCO_0, 1 -> LOCO_1, 2 -> LOCO_2`, and the model predicts the class `1` on this row. Then we want to return `LOCO_1`

Reply:
let's add some docs to explain it, cause it looks weird to me.